% \VignetteIndexEntry{tcgaR: }
% \VignettePackage{tcgaR}
% \VignetteDepends{BiocStyle}
% \VignetteEngine{knitr::knitr}

\documentclass[12pt]{article}
\usepackage[numbers]{natbib}
\usepackage{amsthm}
\usepackage{amssymb}
\newcommand{\minfi}{\Biocpkg{minfi}}

<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

\title{tcgaR: }
\author{Jean-Philippe Fortin, Kasper Daniel Hansen}
\begin{document}

\maketitle{}
\setcounter{secnumdepth}{1} 

% Introduction
\section{Introduction}

Up to now, more than 10,000 methylation samples from the state-of-the-art 450K microarray have been made available through The Cancer Genome Atlas portal \citep{tcga} and the Gene Expression
Omnibus (GEO) \citep{Geo}. Large-scale comparison studies, for instance between cancers or tissues,
become possible epigenome-widely. These large studies often require a substantial amount of time
spent on preprocessing the data and performing quality control. For such
studies, it is not rare to encounter significant batch effects, and those can have a dramatic impact
on the validity of the biological results \citep{batchreview,Harper:2013}. With that in mind, we developed \Biocpkg{shinyMethyl} to make the preprocessing of large 450K datasets intuitive,
enjoyable and reproducible.  \Biocpkg{shinyMethyl} is an interactive visualization tool for Illumina
450K methylation array data based on the packages \Biocpkg{minfi} and \CRANpkg{shiny}
\citep{minfi,shiny}.

\bioccomment{Some comment} 

\section{Data}

Data: April 6 2015

\begin{table}[!ht]
\begin{center}
    \begin{tabular}{ c|l|c|c}
  Abb. & Full name & 27k & 450k  \\
  \hline
  ACC & Adrenocortical carcinoma && \checkmark \\
  BLCA & Bladder Urothelial Carcinoma && \checkmark\\
  BRCA & Breast invasive carcinoma & \checkmark& \checkmark\\
  CESC & Cervical squamous cell carcinoma && \checkmark\\
  CHOL & Cholangiocarcinoma && \checkmark\\
COAD & Colon adenocarcinoma & \checkmark& \checkmark\\
DLBC & Large B-cell Lymphoma && \checkmark\\
ESCA & Esophageal carcinoma && \checkmark \\
GBM & Glioblastoma multiforme & \checkmark& \checkmark\\
HNSC & Head and Neck squamous cell carcinoma  && \checkmark\\
KICH & Kidney Chromophobe&& \checkmark\\
KIRP& Kidney renal papillary cell carcinoma& \checkmark& \checkmark\\
KIRC &Kidney renal clear cell carcinoma& \checkmark& \checkmark\\
LAML & Acute Myeloid Leukemia & \checkmark& \checkmark\\
LCML & Chronic Myelogenous Leukemia & & \\
LGG & Lower grade glioma && \checkmark \\
LIHC & Liver hepatocellular carcinoma && \checkmark\\
LUAD & Lung adenocarcinoma& \checkmark& \checkmark\\
LUSC & Lung squamous cell carcinoma& \checkmark& \checkmark\\
MESO & Mesothelioma && \checkmark\\
OV & Ovarian serous cystadenocarcinoma& \checkmark& \checkmark\\
PAAD & Pancreatic adenocarcinoma && \checkmark\\
PCPG & Pheochromocytoma and Paraganglioma && \checkmark\\
PRAD & Prostate adenocarcinoma&& \checkmark \\
READ & Rectum adenocarcinoma& \checkmark& \checkmark\\
SARC & Sarcoma && \checkmark\\
SKCM & Skin Cutaneous Melanoma && \checkmark\\
STAD & Stomach adenocarcinoma& \checkmark& \checkmark\\
TGCT & Testicular Germ Cell Tumors && \checkmark\\
THCA & Thyroid carcinoma&& \checkmark \\
THYM & Thymoma&& \checkmark  \\
UCEC & Uterine Corpus Endometrial Carcinoma & \checkmark& \checkmark\\
UCS & Uterine Carcinosarcoma && \checkmark\\
UVM & Uveal Melanoma && \checkmark
    \end{tabular}
\end{center}
\caption{Hi-C data sources.}\label{tab:hicdata}
\end{table}

\section*{DNA Methylation 27k data}


\begin{table}[!ht]
\begin{center}
    \begin{tabular}{ c|l|c|c}
  Abb. & Full name & N Normal & N Tumor \\
  \hline
  BRCA & Breast invasive carcinoma & 27 & 318\\
COAD & Colon adenocarcinoma & 37 & 166\\
GBM & Glioblastoma multiforme & 0 & 296\\
KIRC &Kidney renal clear cell carcinoma & 199 & 219\\
KIRP& Kidney renal papillary cell carcinoma&  5 & 16 \\
LAML & Acute Myeloid Leukemia & 0 & 194\\
LUAD & Lung adenocarcinoma& 24 & 127 \\
LUSC & Lung squamous cell carcinoma& 27 & 134\\
OV & Ovarian serous cystadenocarcinoma& 14 & 603\\
READ & Rectum adenocarcinoma& 5 & 68\\
STAD & Stomach adenocarcinoma& 59 & 82\\
UCEC & Uterine Corpus Endometrial Carcinoma & 1 & 117\\
    \end{tabular}
\end{center}
\caption{Hi-C data sources.}\label{tab:hicdata}
\end{table}



\subsection*{Get methylation mappings}

The function \Rfunction{getMethMappings} allows to get all the mappings. 

\section*{Clinical data}

The function \Rfunction{getClinicalData} allows to get all the available data for each cancer type. 

\bioccomment{some comment}

\section*{Session info}

Here is the output of \Rfunction{sessionInfo} on the system on which
this document was compiled:
%<<sessionInfo, results="asis", echo=FALSE, eval=TRUE>>=
%toLatex(sessionInfo())
%@


\bibliography{tcgaR}

\newpage







%\begin{figure}[htbp]
%\begin{center}
%\includegraphics[width=0.9\textwidth]{figures/BothStatus.pdf}
%\caption{\textbf{Title}}
%\label{status}
%\end{center}
%\end{figure}

\end{document}

